docs/proposals/20200720-single-controller-multitenancy.md

```diff
@@ -0,0 +1,559 @@
+
+```go
+// AzureClients contains all the Azure clients used by the scopes.
+type AzureClients struct {
+	SubscriptionID             string
+	ResourceManagerEndpoint    string
+	ResourceManagerVMDNSSuffix string
+	Authorizer                 autorest.Authorizer
+}
+
+```
+
+The signatures for the functions which create these instances are as follows:
+
+```go
+func NewClusterScope(params ClusterScopeParams) (*ClusterScope, error) {
+  ...
+  return &ClusterScope{
+    ...
+  }, nil
+}
+```
+
+#### Proposed Changes
+The proposed changes below only apply to infrastructure nodes which run the CAPZ controllers.
+These changes have no impact on the identities specified on AzureMachines or other workload 
+infrastructure.
+
+##### AAD Pod Identity for Attaching Managed Identities
+[AAD Pod Identity](https://github.com/Azure/aad-pod-identity) enables Kubernetes applications to 
+access cloud resources securely using AAD Managed Identities. To enable CAPZ to authenticate as an 
+Azure Managed Identity, either System Assigned or User Assigned Identities, CAPZ needs to have 
+access to the Azure IMDS service running on the host. This can be accomplished indirectly by using 
+AAD Pod Identity.
+
+With AAD Pod Identity running within the management cluster CAPZ can create [AzureIdentityBindings](https://github.com/Azure/aad-pod-identity#5-deploy-azureidentitybinding)
+and other related structures to enable CAPZ to bind to Azure Managed Identities.
+
+To use Azure Managed Identities, the infrastructure nodes hosting the CAPZ controller must be hosted
+in Azure. Outside of Azure, Service Principal identities are the only available identity type.
+
+##### Cluster API Provider Azure v1alpha3 types
+
+<strong><em>Changed Resources</strong></em>
+* `AzureCluster`
+
+<strong><em>New Resources</strong></em>
+
+<em>Cluster scoped resources</em>
+
+* `AzureServicePrincipal` represents an AAD Service Principal. Should support both 
+   certificate and text secrets.
+* `AzureSystemAssignedIdentity` represents an Azure System Assigned Identity provided 
+   [via Pod Identity](#aad-pod-identity-for-attaching-managed-identities)
+* `AzureUserAssignedIdentity` represents an Azure User Assigned Identity provided 
+   [via Pod Identity](#aad-pod-identity-for-attaching-managed-identities)
+
+<strong><em>Changes to AzureCluster</em></strong>
+
+A new field is added to the `AzureClusterSpec` to reference a principal. We intend to use 
+`corev1.LocalObjectReference` in order to ensure that the only objects that can be references are 
+either in the same namespace or are scoped to the entire cluster.
+
+```go
+// AzurePrincipalKind defines allowed Azure principal types
+// +kubebuilder:validation:Enum=AzureSystemAssigned;AzureUserAssigned;AzureServicePrincipal
+type AzurePrincipalKind string
+
+type AzurePrincipalRef struct {
+  Kind AzurePrincipalKind `json:"kind"`
+  Name string `json:"name"`
+}
+
+type  AzureClusterSpec  struct {
```

Comment:
I just thought about this, what's the impact on the subscriptionID field in AzureClusterSpec? None right?

Reply:
The changes should have no impact on subscriptionID. 

An AAD identity has nothing to do with a subscription. Auth only needs the tenant, client ID, a secret, and resource or scopes.

If one day in the future we decide to build multi-subscription clusters, then we'd have to make a change to subscription ID.

Reply:
Got it. Thanks for the clarification.